Prevent editor clock from continuing beyond song end

voidedWarranties opened the pull request:
Fixes #8134

Behavior now matches stable

Commit:
Prevent playback from going beyond song end

bdach commented on osu.Game/Screens/Edit/EditorClock.cs:
```diff
@@ -37,6 +39,23 @@ public EditorClock(ControlPointInfo controlPointInfo, double trackLength, Bindab
+                }
+            }
+            else
+                PlaybackFinished = false;
```
This is super weirdly written. If the track is running then `false` gets assigned every frame, but when it finishes `true` gets assigned only once. It feels (a) way too complex (if it's not a bottleneck, and I'd be very surprised if it was, it should just assign every frame), and (b) if we're even doing microoptimisations like this then it should probably be the other way around since the `true` path is the (*very*) unlikely one.

Commit:
Internalize both looping and stopping

bdach reviewed (changes requested):
This has non-trivial behaviour (editor clock stops upon reaching end and then restarts when it is sought to end and re-started), and as such it should probably be tested to avoid it regressing in the future.

Commit:
Test stopping behavior

bdach reviewed (dismissed):
I think this is fine for now 👍

smoogipoo commented:
This seems to behave weirdly in the actual editor. When manually seeking to the end while the track is playing, it can restart the track: https://drive.google.com/file/d/1dzD3UM8JTGvcMvkAS6OaC464AJXEJNgK

Commit:
Fix seeking back to beginning too early

voidedWarranties commented:
I've fixed this behavior, and also prevented the `Timeline` from forcing the clock to seek past the end of the track.

peppy committed:
Merge branch 'master' into editor-song-end

peppy commented:
This needs proper testing (hopefully via a test scene with asserts). It's more broken than before, I think:

![image](https://user-images.githubusercontent.com/191335/77501554-c017e880-6e9b-11ea-80e5-7868ec769681.gif)

voidedWarranties commented:
I think the preview track looping feature is not correctly disabled when accessing the editor via the context songselect menu. (Most of) the brokenness shown above is present in the latest release.
This behavior works correctly when used via the main menu.

peppy commented:
I see. Please open an issue for that if you can. Would still appreciate test coverage.

peppy committed:
Merge branch 'fix-song-select-looping' into editor-song-end

peppy committed:
Better ensure track restarted

peppy commented:
As an aside, the new behaviour does not match stable. stable makes specific allowances for storyboarding outside of the track end time:

![image](https://user-images.githubusercontent.com/191335/77510586-80a8c680-6eb2-11ea-8381-1c37e09c097d.gif)

peppy commented:
This PR is approaching the issue in the wrong way. To stop the editor playing beyond the end (which was intentionally done this way to test behaviour we will need in the future (see above example for storyboard editor)) the editor clock just needs to be set to coupled mode.

https://github.com/ppy/osu/blob/ebe3cf296640971fb9d95be0fc76faa1be548545/osu.Game/Screens/Edit/Editor.cs#L81

The only remaining issue after setting this (and fixing a framework bug with coupled seeking) is to decide whether to restart the track or continue playing.

peppy commented:
One concern of using `IsCoupled = true` is that all seek operations will be blocking on async bass operations, for a `TrackBass`. This may be an oversight or may be something we have to live with. It is one reason to potentially leave the clock decoupled, but at this point it defeats the purpose of having the `IsCoupled` toggle in the first place (it was made with the editor in mind).

bdach committed:
Merge branch 'master' into editor-song-end

bdach committed:
Adjust test namespace

bdach committed:
Fix test failure by setting beatmap

Post-merge, it was failing because somewhere along the way
`EditorClockTestScene` started expecting inheritors to set the beatmap
themselves explicitly in BDL.

bdach committed:
Fix test hangs in browser due to changing tracks via music controller

bdach commented:
I have attempted necromancy on this branch to have it available without the merge conflicts if we choose to work on top of it, as it seems to be the better than the other one PR'd today.

That said, I'm not a huge fan of this branch's logic, and the following test that I'd expect to pass fails on this branch:

```csharp
        [Test]
        public void TestWrapWhenStoppedAtTrackEnd()
        {
            AddStep("reset clock", () => Clock.Seek(0));

            AddStep("stop clock", Clock.Stop);
            AddAssert("clock running", () => !Clock.IsRunning);

            AddStep("seek exactly to end", () => Clock.Seek(Clock.TrackLength));
            AddAssert("clock stopped at end", () => Clock.CurrentTime == Clock.TrackLength);

            AddStep("start clock again", Clock.Start);
            AddAssert("clock looped to start", () => Clock.IsRunning && Clock.CurrentTime < 500);
        }
```

Admittedly I'm not very familiar with the woes of the decoupleable clock, so I'm not sure whether I can make that target solution work. @peppy do you perhaps recall what the problem was there?

voidedWarranties committed:
Improve logic and add previously failing test

Co-authored-by: Bartłomiej Dach <dach.bartlomiej@gmail.com>

voidedWarranties commented:
The most recent commit moves the looping logic to `Start`, which I think is a bit less sketchy.
It also fixes the failing test above.

bdach commented on osu.Game/Screens/Edit/EditorClock.cs:
```diff
@@ -216,7 +222,21 @@ public void SeekSmoothlyTo(double seekDestination)
         public bool IsRunning => underlyingClock.IsRunning;
 
-        public void ProcessFrame() => underlyingClock.ProcessFrame();
+        public void ProcessFrame()
+        {
+            underlyingClock.ProcessFrame();
+
+            var playbackAlreadyStopped = playbackFinished;
+            playbackFinished = CurrentTime >= TrackLength;
+
+            if (playbackFinished && !playbackAlreadyStopped)
+            {
+                if (IsRunning)
+                    underlyingClock.Stop();
+
+                underlyingClock.Seek(TrackLength);
+            }
+        }
```
I'm thinking that this could be simplified further to

```csharp
        public void ProcessFrame()
        {
            underlyingClock.ProcessFrame();

            playbackFinished = CurrentTime >= TrackLength;

            if (playbackFinished)
            {
                if (IsRunning)
                    underlyingClock.Stop();

                if (CurrentTime > TrackLength)
                    underlyingClock.Seek(TrackLength);
            }
        }
```

Reasoning for why:
* Gets rid of another conditional variable
* Ensures that the time always is rolled back to `TrackLength` (I can imagine a scenario where a clock seek stops at `TrackLength` first, but then, because in the current code we have `playbackAlreadyStopped`, you could technically seek further past the track length, as that variable being `true` makes the code inside dead).

Thoughts?

voidedWarranties replied:
Seems like a no-brainer to me

voidedWarranties committed:
Further simplify logic

peppy commented:
@bdach I'd have to double check, but I think it has to do something with the editor doing `Transform` logic on the time of the clock, and if it was changed to be directly coupled, each frame would trigger a seek on the bass track, causing a considerable overhead.

In fact, I remember this being a thing at one point, so the decoupling may actually be turned on as a "workaround" for that..

peppy commented:
As a temporary change I think this looks fine. There's still a larger piece of the puzzle remaining, which is sharing offset/clock logic with `GameplayClock` (see #11400) at which point this logic may need to be removed or replaced.

Let's get this in for now.

peppy commented on osu.Game.Tests/Visual/Editing/TestSceneEditorClock.cs:
```diff
@@ -0,0 +1,81 @@
+            AddAssert("clock running", () => Clock.IsRunning);
+
+            AddStep("seek near end", () => Clock.Seek(Clock.TrackLength - 250));
+            AddUntilStep("clock stops", () => !Clock.IsRunning);
```
Super not confident this is going to hold up on CI, but I guess we'll see in time. I can't repro a failure locally.

peppy commented:
While testing this I'm 101% sure there are going to be further changes required, but I will PR those as follow-ups.
